src/index.ts

```diff
@@ -15,62 +25,115 @@ const buildRestClient = (apiClient: InfisicalApi, requestOptions?: RawAxiosReque
-type InfisicalSDKOptions = {
-	siteUrl?: string;
-};
-
 class InfisicalSDK {
 	#apiInstance: InfisicalApi;
 
+	sdkOptions: InfisicalSDKOptions;
+
 	#requestOptions: RawAxiosRequestConfig | undefined;
 	#secretsClient: SecretsClient;
 	#dynamicSecretsClient: DynamicSecretsClient;
 	#authClient: AuthClient;
-	#basePath: string;
+	#tokenDetails?: TTokenDetails;
+	#authCredentials?: TAuthCredentials;
 
 	constructor(options?: InfisicalSDKOptions) {
-		this.#basePath = options?.siteUrl || "https://app.infisical.com";
+		this.sdkOptions = {
+			autoTokenRefresh: options?.autoTokenRefresh ?? true,
+			siteUrl: options?.siteUrl || "https://app.infisical.com"
+		};
 
 		this.#apiInstance = new InfisicalApi(
 			new Configuration({
-				basePath: this.#basePath
+				basePath: this.sdkOptions.siteUrl
 			})
 		);
 
-		this.#authClient = new AuthClient(this.authenticate.bind(this), this.#apiInstance, this.#basePath);
+		// We need to do bind(this) because the authenticate method is a private method, and usually you can't call private methods from outside the class.
+		this.#authClient = new AuthClient(this.#authenticator, this.#apiInstance, this.#tokenDetails?.accessToken);
 		this.#dynamicSecretsClient = new DynamicSecretsClient(this.#apiInstance, this.#requestOptions);
 		this.#secretsClient = new SecretsClient(this.#apiInstance, this.#requestOptions);
 		this.rest = () => buildRestClient(this.#apiInstance, this.#requestOptions);
+
+		if (this.sdkOptions.autoTokenRefresh) {
+			this.handleTokenLifecycle();
+		}
 	}
 
-	private authenticate(accessToken: string) {
-		this.#apiInstance = new InfisicalApi(
-			new Configuration({
-				basePath: this.#basePath,
-				accessToken
-			})
-		);
+	private handleTokenLifecycle = async () => {
+		const authMethodMap: Record<AuthMethod, () => Promise<InfisicalSDK>> = {
+			[AuthMethod.UniversalAuth]: async () =>
+				await this.#authClient.universalAuth.login(this.#authCredentials?.credentials as UniversalAuthCredentials),
+			[AuthMethod.AWSIam]: async () => await this.#authClient.awsIamAuth.login(this.#authCredentials?.credentials as AWSIamCredentials),
+			[AuthMethod.AccessToken]: async () =>
+				await this.#authClient.accessToken((this.#authCredentials?.credentials as AccessTokenCredentials).accessToken)
+		};
+
+		while (true) {
+			try {
+				await new Promise(resolve => setTimeout(resolve, 10_000));
+				if (!this.#tokenDetails || !this.#authCredentials) {
+					continue;
+				}
+
+				const now = new Date();
+				const timeSinceFetchInSeconds = (now.getTime() - this.#tokenDetails.fetchedTime.getTime()) / 1_000;
+				const firstFetchTimeInSeconds = (now.getTime() - this.#tokenDetails.firstFetchTime.getTime()) / 1_000;
 
-		this.#requestOptions = {
-			headers: {
-				Authorization: `Bearer ${accessToken}`
+				// If there's less than 15 seconds until the token expires, we reauthenticate early
+				if (firstFetchTimeInSeconds >= this.#tokenDetails.accessTokenMaxTTL - 15) {
```

Comment:
So this would mean that if the ttl was 15 or less, it would keep renewing indefinitely? How about we do a percent, like 2/3rds of the way to end of life and then we renew

Reply:
It might also be worth putting a warning if the max access token/access token ttl is shorter than how much we are subtracting because in those cases it will keep re logging in and firing tons of API calls